test/Dialect/FIRRTL/errors.mlir

```diff
@@ -871,3 +871,115 @@ firrtl.circuit "Foo"   {
     %bar_a, %bar_b, %bar_c = firrtl.instance bar sym [<@w3,1,public>,<@w3,2,private>,<@syh2,0,public>] @Bar(in a: !firrtl.uint<1> [{one}], out b: !firrtl.bundle<baz: uint<1>, qux: uint<1>> [{circt.fieldID = 1 : i32, two}], out c: !firrtl.uint<1> [{four}])
   }
 }
+
+// -----
+// Check single usage of reference ports
+firrtl.circuit "Foo" {
+  firrtl.module @Bar(in %_a: !firrtl.ref<uint<1>>) {
+    %a = firrtl.wire : !firrtl.uint<1>
+    firrtl.ref.recv %a, %_a : !firrtl.ref<uint<1>>
+  }
+  firrtl.module @Foo(in %fa: !firrtl.ref<uint<1>>) {
+    %bar_a = firrtl.instance bar @Bar(in _a: !firrtl.ref<uint<1>>)
+
+    %zero = firrtl.constant 0 : !firrtl.uint<1>
+    // expected-error @+1 {{reference port operand cannot be reused by any other op}}
+    %0 = firrtl.ref.resolve %bar_a : !firrtl.ref<uint<1>>
+    firrtl.strictconnect %0, %zero : !firrtl.uint<1>
+    firrtl.strictconnect %fa, %bar_a : !firrtl.ref<uint<1>>
+
+  }
+}
+
+// -----
+// Check upward reference XMRs
+
+firrtl.circuit "MyView_mapping" {
+  firrtl.module @MyView_mapping(in %ref_in1: !firrtl.ref<uint<1>>) {
+    // expected-error @+1 {{'firrtl.ref.resolve' op failed to verify that it cannot have module port as an operand because it implies upward reference XMR, which are not supported. The reference operand must be a port from firrtl.instance}}
+    %0 = firrtl.ref.resolve %ref_in1 : !firrtl.ref<uint<1>>
+  }
+}
+
+// -----
+// Incorrect connections
+
+firrtl.circuit "MyView" {
+  firrtl.module @MyView() {
+    %ref_in1 = firrtl.wire : !firrtl.ref<uint<1>>
+    %in1 = firrtl.wire : !firrtl.uint<1>
+    // expected-error @+1 {{that reference port operand can only be a module or instance port}}
+    firrtl.ref.send  %ref_in1, %in1 : !firrtl.ref<uint<1>>
+  }
+}
+
+// -----
+// Check upward reference XMRs
+
+firrtl.circuit "func2" {
+  firrtl.module @func2(in %ref_in1: !firrtl.ref<uint<1>>, out %ref_out: !firrtl.ref<uint<1>>) {
+    // expected-error @+1 {{connect has invalid flow for Ref type ports: the source expression "ref_in1" and destination expression "ref_out" both have same port kind, expected Module port to Instance connections only}}
+    firrtl.strictconnect %ref_out, %ref_in1 : !firrtl.ref<uint<1>>
+  }
+}
+
+// -----
+// Check flow semantics for ref.recv
+
+firrtl.circuit "Foo" {
+  // expected-note @+1 {{source was defined here}}
+  firrtl.module @Foo(out %_a: !firrtl.ref<uint<1>>) {
+    %a = firrtl.wire : !firrtl.uint<1>
+    // expected-error @+1 {{connect has invalid flow: the source expression "_a" has sink flow, expected source or duplex flow}}
+    firrtl.ref.recv %a, %_a : !firrtl.ref<uint<1>>
```

Comment:
This check is technically incorrect. `%a` cannot be duplex here due to constraints on references not being passive. Ditto for `ref.send` below.